packages/kbn-test/types/ftr.d.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ToolingLog } from '@kbn/dev-utils';
+import { DefaultServiceProviders } from '../../../src/functional_test_runner/types';
+
+type Provider = (ctx: GenericFtrProviderContext<any, any>) => any;
+type ProviderMap = Record<string, Provider>;
+
+interface AsyncInstance<T> {
+  /**
+   * Services that are initialized async are not ready before the tests execute, so you might need
+   * to call `init()` and await the promise it returns before interacting with the service
+   */
+  init(): Promise<T>;
+}
+
+/**
+ * When a provider returns a promise it is initialized as an AsyncInstance that is a
+ * proxy to the eventual result with an added init() method which returns the eventual
+ * result. Automatically unwrap these promises and convert them to AsyncInstances + Instance
+ * types.
+ */
+type MaybeAsyncInstance<T> = T extends Promise<infer X> ? AsyncInstance<X> & X : T;
+
+/**
+ * Convert a map of providers to a map of the instance types they provide, also converting
+ * promise types into the async instances that other providers will receive.
+ */
+type ProvidedTypeMap<T extends ProviderMap> = {
+  [K in keyof T]: MaybeAsyncInstance<ReturnType<T[K]>>
+};
+
+export interface GenericFtrProviderContext<
```

Comment:
> Describe your solution about the interface with 4! generics and 2 of them are already defined

Yeah, 4 is a lot but I mostly see it as a generic interface with two arguments.

> maybe you probably need to use `type` instead of super-generic interface?

Hmm, I would if I could but I don't think there's a valid way to derive a type inside the interface except like this. Something like this would be ideal:

```ts
export interface GenericFtrProviderContext<
  ServiceProviders extends object,
  PageObjectProviders extends object
> {
  type ServiceMap = ProvidedTypeMap<ServiceProviders & DefaultServiceProviders>
  type PageObjectMap = ProvidedTypeMap<PageObjectProviders>

  ...
}
```
But that's not valid TS, and I use both of those named types twice, and would prefer to not have to repeat their initializers in both places. Using defined generic arguments is aimed at making it more readable.

Reply:
you can use like this😊:
``` ts
type ServiceMap = ProvidedTypeMap<ProviderMap & DefaultServiceProviders>;
type PageObjectMap = ProvidedTypeMap<ProviderMap>;
export interface GenericFtrProviderContext<
  ServiceProviders extends ProviderMap,
  PageObjectProviders extends ProviderMap
> {/* code here */}
```